core/src/value/partial.rs

```diff
@@ -0,0 +1,873 @@
+use snafu::{Backtrace, OptionExt, ResultExt, Snafu};
+use std::convert::{TryFrom, TryInto};
+use std::fmt;
+use std::ops::RangeInclusive;
+
+#[derive(Debug, Snafu)]
+#[non_exhaustive]
+pub enum Error {
+    #[snafu(display("Date is invalid."))]
+    InvalidDate { backtrace: Backtrace },
+    #[snafu(display("Time is invalid."))]
+    InvalidTime { backtrace: Backtrace },
+    #[snafu(display("DateTime is invalid."))]
+    InvalidDateTime { backtrace: Backtrace },
+    #[snafu(display("To combine a DicomDate with a DicomTime value, the DicomDate has to be precise. Precision is: '{:?}'.", value))]
+    DateTimeFromPartials {
+        value: DateComponent,
+        backtrace: Backtrace,
+    },
+    #[snafu(display(
+        "'{:?}' has invalid value: '{}', must be in {:?}",
+        component,
+        value,
+        range
+    ))]
+    InvalidComponent {
+        component: DateComponent,
+        value: u32,
+        range: RangeInclusive<u32>,
+        backtrace: Backtrace,
+    },
+    #[snafu(display(
+        "Second fraction precision '{}' is out of range, must be in 0..=6.",
+        value
+    ))]
+    FractionPrecisionRange { value: u32, backtrace: Backtrace },
+    #[snafu(display(
+        "Number of digits in decimal representation of fraction '{}' does not match it's precision '{}'.",
+        fraction,
+        precision
+    ))]
+    FractionPrecisionMismatch {
+        fraction: u32,
+        precision: u32,
+        backtrace: Backtrace,
+    },
+    #[snafu(display("Conversion Failed."))]
+    Conversion { source: std::num::TryFromIntError },
+    #[snafu(display(
+        "Cannot convert from an imprecise value. This value represents a date / time range."
+    ))]
+    ImpreciseValue { backtrace: Backtrace },
+}
+
+type Result<T, E = Error> = std::result::Result<T, E>;
+
+/// Represents components of Date, Time and DateTime values.
+#[derive(Debug, PartialEq)]
+pub enum DateComponent {
+    Year,
+    Month,
+    Day,
+    Hour,
+    Minute,
+    Second,
+    Fraction,
+    UTCOffset,
+}
+
+/// Represents a Dicom Date value with a partial precision,
+/// where some date components may be missing.
+///
+/// Unlike RUSTs `chrono::NaiveDate`, it does not allow for negative years.
+///
+/// Although the DICOM protocol does not allow for an incomplete DA value,
+/// this structure is necessary for range matching purposes, where incomplete
+/// date values occur.
+#[derive(Debug, Clone, Copy, PartialEq)]
+pub enum DicomDate {
+    Year(u16),
+    Month(u16, u8),
+    Day(u16, u8, u8),
+}
+
+/// Represents a Dicom Time value with a partial precision,
+/// where some time components may be missing.
+///
+/// Unlike RUSTs `chrono::NaiveTime`, this implemenation of time is DICOM compliant:  
+/// - has only 6 digit precision for fraction of a second
+/// - has no means to handle leap seconds
+#[derive(Debug, Clone, Copy, PartialEq)]
+pub enum DicomTime {
```

Comment:
`DicomTime` and `DicomDate` have publicly visible variants, which can let users create inconsistent values. I would strongly recommend employing a newtype to hide the enum-based implementation.

```rust
#[derive(Debug, Clone, Copy, PartialEq)]
pub struct DicomTime(DicomTimeImpl);

#[derive(Debug, Clone, Copy, PartialEq)]
enum DicomTimeImpl {
    // ...
}
```